common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java

```diff
@@ -176,15 +176,31 @@ public Collation(
    */
 
   public static StringSearch getStringSearch(
-      final UTF8String left,
-      final UTF8String right,
+      final UTF8String targetUTF8String,
+      final UTF8String patternUTF8String,
       final int collationId) {
-    String pattern = right.toString();
-    CharacterIterator target = new StringCharacterIterator(left.toString());
+
+    if (collationId == UTF8_BINARY_COLLATION_ID) {
+      return getStringSearch(targetUTF8String, patternUTF8String);
```

Comment:
Why would we ever do this against UTF8_BINARY? For UTF8_BINARY we should just stay on binary level and avoid string copy.

Reply:
Following [comment1](https://github.com/apache/spark/pull/45791#discussion_r1549312876) and [comment2](https://github.com/apache/spark/pull/45791#discussion_r1551446372) we decided to do this just in case because this is public method and all collationId values should be handled. We never call this method with `collationId == 0`.

Reply:
Then can we just add an assert? This is not public and only used by Spark internally.

Reply:
I think assert would do the job for UTF8_BINARY. That way we know not to call this method for UTF8_BINARY collation, and there will be no unnecessary/confusing code